File: tsort.rb
#!/usr/bin/env ruby

# topological sort with detection
# http://en.wikipedia.org/wiki/Topological_sorting

def visit(graph, node, deps, visited, &block)
  v = visited[node]
  if v == :just
    return true
  elsif v == :once
    return false
  else
    visited[node] = :just
    deps.each do |dep|
      if visit(graph, dep, graph[dep], visited, &block)
        return true
      end
    end
    visited[node] = :once
    block.call(node)
  end
end

def tsort(graph, &block)
  visited = {}
  
  graph.each_pair do |node, deps|
    return true if visit(graph, node, deps, visited, &block)
  end
  
  return false
end

test_data = {
  :graph1 => {
    'a' => ['b'],
    'b' => ['c'],
    'c' => [],
  },

  :graph2 => {
    'a' => [],
    'b' => [],
    'c' => ['a', 'b'],
  },

  :graph3 => {
    'a' => ['a'],
  },

  :graph4 => {
    'a' => ['b'],
    'b' => ['c'],
    'c' => ['a'],
  },

  :graph5 => {
    'a' => ['b', 'c'],
    'b' => ['a', 'c'],
    'c' => ['a', 'b'],
  },

}

test_data.each_pair do |name, graph|
  puts "#{name.to_s}"
  print "  in  : "; p graph

  print "  out : "
  cyclic = tsort(graph) do |node|
    print "#{node} -> "
  end
  print "\n"
  puts cyclic ? "  cyclic" : "  asyclic"
end

File: tsort.lua

function tsort(graph, on_resolved)
	visited = {}
	
	function visit(node, deps)
		v = visited[node]
		if v == "just" then
			return true
		elseif v == "once" then
			return false
		else
			visited[node] = "just"
			for k, v in pairs(deps) do
				if visit(v, graph[v]) then
					return true
				end
			end
			visited[node] = "once"
			on_resolved(node)
			return false
		end
	end
	
	for node, deps in pairs(graph) do
		if visit(node, deps) then
			return true
		end
	end
	
	return false
end

graphs = {
	graph1 = {
		lib1 = {'lib2'},
		lib2 = {'lib3'},
		lib3 = {},
	},
	
	graph2 = {
		lib1 = {'lib2'},
		lib2 = {'lib3'},
		lib3 = {'lib1'},
	}
}

for k, v in pairs(graphs) do
	print(k, tsort(v, print))
end
